src/main/java/graphql/execution/instrumentation/dataloader/FieldLevelTrackingApproach.java

```diff
@@ -30,65 +29,67 @@
     private final Supplier<DataLoaderRegistry> dataLoaderRegistrySupplier;
     private final Logger log;
 
-    private static class CallStack implements InstrumentationState {
+    private static class FieldTracking {
 
-        private final Map<Integer, Integer> expectedFetchCountPerLevel = new LinkedHashMap<>();
-        private final Map<Integer, Integer> fetchCountPerLevel = new LinkedHashMap<>();
-        private final Map<Integer, Integer> expectedStrategyCallsPerLevel = new LinkedHashMap<>();
-        private final Map<Integer, Integer> happenedStrategyCallsPerLevel = new LinkedHashMap<>();
-        private final Map<Integer, Integer> happenedOnFieldValueCallsPerLevel = new LinkedHashMap<>();
+        final AtomicInteger expectedFetchCountPerLevel = new AtomicInteger();
+        final AtomicInteger fetchCountPerLevel = new AtomicInteger();
+        final AtomicInteger expectedStrategyCallsPerLevel = new AtomicInteger();
+        final AtomicInteger happenedStrategyCallsPerLevel = new AtomicInteger();
+        final AtomicInteger happenedOnFieldValueCallsPerLevel = new AtomicInteger();
+    }
 
+    private static class CallStack implements InstrumentationState {
 
-        private final Set<Integer> dispatchedLevels = new LinkedHashSet<>();
+        private final ConcurrentMap<Integer, FieldTracking> fieldTrackingPerLevel = new ConcurrentHashMap<>();
+        private final ConcurrentHashMap.KeySetView<Integer, Boolean> dispatchedLevels = ConcurrentHashMap.newKeySet();
 
         CallStack() {
-            expectedStrategyCallsPerLevel.put(1, 1);
+            FieldTracking fieldTracking = new FieldTracking();
+            fieldTracking.expectedStrategyCallsPerLevel.incrementAndGet();
+            fieldTrackingPerLevel.put(1, fieldTracking);
         }
 
+        private FieldTracking provideFieldTracking(int level) {
+            FieldTracking fieldTracking = fieldTrackingPerLevel.get(level);
+            if (fieldTracking == null) {
+                fieldTracking = fieldTrackingPerLevel.computeIfAbsent(level, ignored -> new FieldTracking());
+            }
+            return fieldTracking;
```

Comment:
This can be simplified to 

```
private FieldTracking provideFieldTracking(int level) {
           return  fieldTrackingPerLevel.computeIfAbsent(level, ignored -> new FieldTracking());
}
```

Reply:
here I'm trying to avoid a pitfall of `ConcurrentHashMap`: if the level is already present (cache hit) it is possible to skip some locking of `computeIfAbsent`.

For example also caffeine is using this idiom:
https://github.com/ben-manes/caffeine/blob/6f7dfcf275944ae2c204abe808ae67e035f682a3/caffeine/src/main/java/com/github/benmanes/caffeine/cache/UnboundedLocalCache.java#L224

Reply:
but according to the javadoc:
> Some attempted update operations on this map by other threads may be blocked while computation is in progress, so the computation should be short and simple, and must not attempt to update any other mappings of this map.

computation here `new FieldTracking`, so shouldn't be a problem